### PR TITLE
feat: kashi multi pair lender improvements

### PR DESCRIPTION
**Pull request by fp-crypto**
- feat: don't return loss liquidate can't free loaned assets
- feat: add force param to removeKashiPair
- fix: issue preventing selling rewards immediately
- fix: break out of liquidate loop when `amountToFree` is dust
- fix: don't accrue or claim if a pair is only dust

### CODE REVIEW

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -348,7 +353,9 @@ contract Strategy is BaseStrategy {
                 for (
                     uint256 i = 0;
                     i < kashiPairs.length &&
-                        sharesToFreeFromKashi > sharesFreedFromKashi;
+                        sharesToFreeFromKashi > sharesFreedFromKashi &&
+                        sharesToFreeFromKashi.sub(sharesFreedFromKashi) >
```
Comment (jmonteer) [inline contracts/Strategy.sol]: question: 

does it make sense to use the threshold when withdrawing? 

this may lead to the strategy leaving dust in the strategy / returning less than requested. this should not be a real problem in this case but in a general situation I don't use the threshold for withdrawals. 

only use of threshold for withdrawals I do is to withdraw the full balance if I am leaving dust invested. 

what do you think?

Comment (jmonteer) [reply contracts/Strategy.sol]: or is there any situation in which there are rounding errors and the dust will never be able to be withdrawed?

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -557,10 +572,7 @@ contract Strategy is BaseStrategy {
         if (pid == 0) return;
 
         uint256 fractionsToStake = kashiFractionInPair(kashiPair);
-
-        if (fractionsToStake > dustThreshold) {
```
Comment (jmonteer) [inline contracts/Strategy.sol:561]: why did you remove the threshold here?

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -260,13 +266,12 @@ contract Strategy is BaseStrategy {
...
 
         if (amountToFree > 0 && wantBal < amountToFree) {
-            (uint256 newLoose, ) = liquidatePosition(amountToFree.sub(wantBal));
+            (uint256 newLoose, ) = liquidatePosition(amountToFree);
 
             // if we didnt free enough money, prioritize paying down debt before taking profit
             if (newLoose < amountToFree) {
                 if (newLoose <= _debtPayment) {
                     _profit = 0;
-                    _loss += _debtPayment.sub(newLoose);
```
Comment (jmonteer) [inline contracts/Strategy.sol:269]: what would happen in the case of a real loss? not a locked funds loss but a hack or bug that made kashi LPs to lose money? 

maybe is there a way to find out the reason we've not been able to withdraw full amount? (like checking available liquidity)

File: contracts/Strategy.sol [hunk 4]
```diff
@@ -236,6 +236,12 @@ contract Strategy is BaseStrategy {
     {
         for (uint256 i = 0; i < kashiPairs.length; i++) {
             KashiPairInfo memory kashiPairInfo = kashiPairs[i];
+            if (
+                kashiFractionTotal(
+                    kashiPairInfo.kashiPair,
+                    kashiPairInfo.pid
+                ) <= dustThreshold
+            ) continue; // skip if there's only dust
```
Comment (jmonteer) [inline contracts/Strategy.sol]: minor edge case:

if the strategy assumes to be using accrued kashiPairs, can this be a problem for those KashiPairs that hold dust if they are interacted with in any way?

**Review by jmonteer — COMMENTED** [review]
Looks good

comments related to the use of the threshold. I would use threshold only for deposits. For withdrrawals I would use an inverse threshold (if I am leaving dust in Kashi, I withdraw all)

**Comment by jmonteer** [comment]
also, I dont see a change for this: 

> fix: issue preventing selling rewards immediately

**Comment by fp-crypto** [comment]
> also, I dont see a change for this: 
> 
> 
> 
> > fix: issue preventing selling rewards immediately
> 
> 

It was related to not calling deposit on the masterchef when the amount was 0. We call deposit to claim the sushi rewards. See the second modification here https://github.com/fp-crypto/kashi-strategy/commit/40c1cb78649193a7c78da37f3cdd20085ca045e6

File: contracts/Strategy.sol (hunk 1 above)
Comment (fp-crypto) [reply contracts/Strategy.sol]: The issue i was noticing was that rounding errors would lead us to visit additional pairs and consume extra gas when the value of visiting those pairs was low. I wanted to add an escape hatch to prevent that. I need to think about it more ...

File: contracts/Strategy.sol (hunk 2 above)
Comment (fp-crypto) [reply contracts/Strategy.sol:561]: This was preventing claiming sushi rewards when the fraction was set to 0. Deposit with 0, claims rewards.

File: contracts/Strategy.sol (hunk 3 above)
Comment (fp-crypto) [reply contracts/Strategy.sol:269]: If estimatedTotal assets has gone down relative to our debt, we have lost money. This will be indicated by kashi fraction exchange rate going down, so it should be accounted for.

File: contracts/Strategy.sol (hunk 4 above)
Comment (fp-crypto) [reply contracts/Strategy.sol]: Potentially, but liquidateKashiPair also calls accrue, so it would only lead to underestimating assets and leaving dust. I guess i should just make this `0` instead.

**Comment by fp-crypto** [comment]
> Looks good
> 
> comments related to the use of the threshold. I would use threshold only for deposits. For withdrrawals I would use an inverse threshold (if I am leaving dust in Kashi, I withdraw all)

294c675 attempts to do the inverse dust thresholding, let me know if it looks right.

File: contracts/Strategy.sol (hunk 3 above)
Comment (jmonteer) [reply contracts/Strategy.sol:269]: yeah I agree. I missed previous mention to loss, that you did not remove. all good

**Comment by jmonteer** [comment]
> > Looks good
> > comments related to the use of the threshold. I would use threshold only for deposits. For withdrrawals I would use an inverse threshold (if I am leaving dust in Kashi, I withdraw all)
> 
> [294c675](https://github.com/fp-crypto/kashi-strategy/commit/294c675e06fa38f5ab8a296c5473b5b9f6ed0407) attempts to do the inverse dust thresholding, let me know if it looks right.

it does.

changes LGTM

**Review by poolpitako — APPROVED** [review]
